### ISSUE TITLE
help me!

### ISSUE THREAD

**Issue by Ava-alt** [clangd/clangd#708]
I want C++ Standard Library Completion。 but i generate  compile_commands.json  but can't up my expectations
expectations
![image](https://user-images.githubusercontent.com/79972211/109788849-b1f54f00-7c4a-11eb-84a8-2a6d5607c6a9.png)
compile_commands.json 
![image](https://user-images.githubusercontent.com/79972211/109788617-6e9ae080-7c4a-11eb-9017-b8a93c9189bf.png)

i want know what do next

**Comment by Ava-alt** [comment]
Actual phenomenon
![image](https://user-images.githubusercontent.com/79972211/109789011-db15df80-7c4a-11eb-8c9d-e0dd6eac8073.png)

**Comment by kirillbobyrev** [comment]
Hi! Could you please provide more information about this issue? I don't really understand what

a) you expect
b) actually happens

I can guess it's related to the header name completion but I don't really see what is the difference in what you're doing for the first and the second screenshot. Could you please provide the additional information as well as clangd version you're usinng?

**Comment by Ava-alt** [comment]
yes,I want  the header name completion, but I don't konw what's the step.
 
  I generate the file of  compile_commands.json by the command 
 cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 . and the file but not fulfill  my expectations.

![image](https://user-images.githubusercontent.com/79972211/109898012-8320bc80-7cce-11eb-860c-e99c901772c0.png)

The environment is ubuntu（Linux version 4.15.0-117-generic (buildd@lcy01-amd64-008) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #118-Ubuntu）

**Comment by Ava-alt** [comment]
yes,I want the header name completion, but I don't konw what's the step.

![image](https://user-images.githubusercontent.com/79972211/109899692-338fc000-7cd1-11eb-80c1-4698c1f3a7a9.png)

I generate the file of compile_commands.json by the command
cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 .  but not fulfill my expectations.


The environment is ubuntu（Linux version 4.15.0-117-generic (buildd@lcy01-amd64-008) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #118-Ubuntu）

**Comment by Ava-alt** [comment]
I[18:00:22.010] clangd version 9.0.0-2~ubuntu18.04.2 (tags/RELEASE_900/final)
I[18:00:22.010] Working directory: /data/code/test11
I[18:00:22.010] argv[0]: /usr/bin/clangd-9
I[18:00:22.010] Starting LSP over stdin/stdout
I[18:00:22.010] <-- initialize(0)
I[18:00:22.011] --> reply:initialize(0) 0 ms
I[18:00:22.011] <-- initialized
I[18:00:22.011] unhandled notification initialized
I[18:00:22.012] <-- textDocument/didOpen
I[18:00:22.012] <-- textDocument/didOpen
I[18:00:22.013] Enqueueing 1 commands for indexing
I[18:00:22.013] --> textDocument/clangd.fileStatus
I[18:00:22.013] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[18:00:22.014] --> textDocument/clangd.fileStatus
I[18:00:22.014] Updating file /data/code/test11/test.h with command inferred from /data/code/test11/helloword.cpp
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -c -x c++-header -std=gnu++14 /data/code/test11/test.h -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[18:00:22.037] --> textDocument/publishDiagnostics
I[18:00:22.037] --> textDocument/clangd.fileStatus
I[18:00:22.586] --> textDocument/publishDiagnostics
I[18:00:22.586] --> textDocument/clangd.fileStatus
I[18:00:24.655] <-- textDocument/didChange
I[18:00:24.655] --> textDocument/clangd.fileStatus
I[18:00:24.831] <-- textDocument/didChange
I[18:00:24.831] --> textDocument/clangd.fileStatus
I[18:00:24.845] <-- textDocument/completion(1)
I[18:00:24.852] Code complete: sema context PreprocessorDirective, query scopes [] (AnyScope=true), expected type <none>
I[18:00:24.852] Code complete: 7 results from Sema, 0 from Index, 0 matched, 0 from identifiers, 7 returned.
I[18:00:24.853] --> reply:textDocument/completion(1) 7 ms
I[18:00:24.999] <-- textDocument/didChange
I[18:00:24.999] --> textDocument/clangd.fileStatus
I[18:00:25.175] <-- textDocument/didChange
I[18:00:25.175] --> textDocument/clangd.fileStatus
I[18:00:25.261] <-- textDocument/didChange
I[18:00:25.261] --> textDocument/clangd.fileStatus
I[18:00:25.502] <-- textDocument/codeAction(2)
I[18:00:25.502] --> textDocument/clangd.fileStatus
I[18:00:25.502] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[18:00:25.866] <-- textDocument/didSave
I[18:00:25.866] unhandled notification textDocument/didSave
I[18:00:25.882] <-- textDocument/didChange
I[18:00:25.882] <-- $/cancelRequest
I[18:00:25.967] --> textDocument/publishDiagnostics
I[18:00:25.967] --> reply:textDocument/codeAction(2) 465 ms, error: Task was cancelled.
I[18:00:25.967] --> textDocument/clangd.fileStatus
[Error - 6:00:25 PM] Request textDocument/codeAction failed.
  Message: Task was cancelled.
  Code: -32001 
I[18:00:26.054] <-- textDocument/didChange
I[18:00:26.054] --> textDocument/clangd.fileStatus
I[18:00:26.303] <-- textDocument/codeAction(3)
I[18:00:26.303] --> textDocument/clangd.fileStatus
I[18:00:26.303] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[18:00:26.324] --> textDocument/publishDiagnostics
I[18:00:26.325] --> reply:textDocument/codeAction(3) 22 ms
I[18:00:26.325] --> textDocument/clangd.fileStatus
I[18:00:26.600] <-- textDocument/codeAction(4)
I[18:00:26.600] --> reply:textDocument/codeAction(4) 0 ms
I[18:00:26.600] --> textDocument/clangd.fileStatus
I[18:00:26.670] <-- textDocument/didSave
I[18:00:26.670] unhandled notification textDocument/didSave
I[18:00:26.689] <-- textDocument/didChange
I[18:00:26.689] --> textDocument/clangd.fileStatus
I[18:00:26.834] <-- textDocument/didChange
I[18:00:26.834] --> textDocument/clangd.fileStatus
I[18:00:27.037] <-- textDocument/didChange
I[18:00:27.037] --> textDocument/clangd.fileStatus
I[18:00:27.155] <-- textDocument/didChange
I[18:00:27.156] --> textDocument/clangd.fileStatus
I[18:00:27.260] <-- textDocument/didChange
I[18:00:27.260] --> textDocument/clangd.fileStatus
I[18:00:27.500] <-- textDocument/codeAction(5)
I[18:00:27.500] --> textDocument/clangd.fileStatus
I[18:00:27.500] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[18:00:27.866] <-- textDocument/didSave
I[18:00:27.866] unhandled notification textDocument/didSave
I[18:00:27.876] --> textDocument/publishDiagnostics
I[18:00:27.876] --> reply:textDocument/codeAction(5) 376 ms
I[18:00:27.876] --> textDocument/clangd.fileStatus
I[18:00:28.150] <-- textDocument/codeAction(6)
I[18:00:28.150] --> reply:textDocument/codeAction(6) 0 ms
I[18:00:28.150] --> textDocument/clangd.fileStatus
I[18:00:28.249] <-- textDocument/didChange
I[18:00:28.250] --> textDocument/clangd.fileStatus
I[18:00:28.492] <-- textDocument/codeAction(7)
I[18:00:28.493] --> textDocument/clangd.fileStatus
I[18:00:28.493] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[18:00:28.811] <-- textDocument/didSave
I[18:00:28.811] unhandled notification textDocument/didSave
I[18:00:28.875] --> textDocument/publishDiagnostics
I[18:00:28.875] --> reply:textDocument/codeAction(7) 382 ms
I[18:00:28.875] --> textDocument/clangd.fileStatus
I[18:00:29.149] <-- textDocument/codeAction(8)
I[18:00:29.149] --> reply:textDocument/codeAction(8) 0 ms
I[18:00:29.149] --> textDocument/clangd.fileStatus
I[09:40:56.084] <-- textDocument/didChange
I[09:40:56.084] --> textDocument/clangd.fileStatus
I[09:40:56.431] <-- textDocument/codeAction(9)
I[09:40:56.431] --> textDocument/clangd.fileStatus
I[09:40:56.431] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[09:40:56.780] --> textDocument/publishDiagnostics
I[09:40:56.780] --> reply:textDocument/codeAction(9) 348 ms
I[09:40:56.780] --> textDocument/clangd.fileStatus
I[09:40:57.453] <-- textDocument/codeAction(10)
I[09:40:57.453] --> reply:textDocument/codeAction(10) 0 ms
I[09:40:57.453] --> textDocument/clangd.fileStatus
I[09:41:24.658] <-- textDocument/didChange
I[09:41:24.658] --> textDocument/clangd.fileStatus
I[09:41:25.158] --> textDocument/clangd.fileStatus
I[09:41:25.158] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[09:41:25.433] <-- textDocument/codeAction(11)
I[09:41:25.583] --> textDocument/publishDiagnostics
I[09:41:25.583] --> reply:textDocument/codeAction(11) 150 ms
I[09:41:25.583] --> textDocument/clangd.fileStatus
I[09:41:25.635] <-- textDocument/didChange
I[09:41:25.635] --> textDocument/clangd.fileStatus
I[09:41:26.135] --> textDocument/clangd.fileStatus
I[09:41:26.135] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[09:41:26.443] <-- textDocument/codeAction(12)
I[09:41:26.484] --> textDocument/publishDiagnostics
I[09:41:26.484] --> reply:textDocument/codeAction(12) 41 ms
I[09:41:26.484] --> textDocument/clangd.fileStatus
I[09:41:27.452] <-- textDocument/codeAction(13)
I[09:41:27.452] --> reply:textDocument/codeAction(13) 0 ms
I[09:41:27.452] --> textDocument/clangd.fileStatus
I[09:41:28.140] <-- textDocument/didChange
I[09:41:28.140] --> textDocument/clangd.fileStatus
I[09:41:28.640] --> textDocument/clangd.fileStatus
I[09:41:28.640] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[09:41:28.989] --> textDocument/publishDiagnostics
I[09:41:28.989] --> textDocument/clangd.fileStatus
I[09:41:29.434] <-- textDocument/codeAction(14)
I[09:41:29.434] --> reply:textDocument/codeAction(14) 0 ms
I[09:41:29.434] --> textDocument/clangd.fileStatus
I[10:14:49.195] <-- textDocument/didChange
I[10:14:49.196] --> textDocument/clangd.fileStatus
I[10:14:49.696] --> textDocument/clangd.fileStatus
I[10:14:49.696] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[10:14:50.121] --> textDocument/publishDiagnostics
I[10:14:50.122] --> textDocument/clangd.fileStatus
I[10:14:50.205] <-- textDocument/didChange
I[10:14:50.206] --> textDocument/clangd.fileStatus
I[10:14:50.399] <-- textDocument/codeAction(15)
I[10:14:50.400] --> textDocument/clangd.fileStatus
I[10:14:50.400] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[10:14:50.785] --> textDocument/publishDiagnostics
I[10:14:50.785] --> reply:textDocument/codeAction(15) 386 ms
I[10:14:50.785] --> textDocument/clangd.fileStatus
I[10:14:51.060] <-- textDocument/codeAction(16)
I[10:14:51.060] --> reply:textDocument/codeAction(16) 0 ms
I[10:14:51.060] --> textDocument/clangd.fileStatus
I[10:14:52.560] <-- textDocument/didChange
I[10:14:52.560] --> textDocument/clangd.fileStatus
I[10:14:53.061] --> textDocument/clangd.fileStatus
I[10:14:53.061] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[10:14:53.583] --> textDocument/publishDiagnostics
I[10:14:53.583] --> textDocument/clangd.fileStatus
I[10:14:53.868] <-- textDocument/codeAction(17)
I[10:14:53.868] --> reply:textDocument/codeAction(17) 0 ms
I[10:14:53.868] --> textDocument/clangd.fileStatus
I[10:14:55.387] <-- textDocument/didChange
I[10:14:55.387] --> textDocument/clangd.fileStatus
I[10:14:55.626] <-- textDocument/codeAction(18)
I[10:14:55.626] --> textDocument/clangd.fileStatus
I[10:14:55.626] Updating file /data/code/test11/helloword.cpp with command 
[/data/code/test11]
/usr/bin/c++ --driver-mode=g++ -I/usr/include/c++/7 -std=gnu++14 -o CMakeFiles/hello.dir/helloword.cpp.o -c /data/code/test11/helloword.cpp -fsyntax-only -resource-dir=/usr/lib/llvm-9/lib/clang/9.0.0
I[10:14:56.025] --> textDocument/publishDiagnostics
I[10:14:56.025] --> reply:textDocument/codeAction(18) 399 ms
I[10:14:56.025] --> textDocument/clangd.fileStatus
I[10:14:56.300] <-- textDocument/codeAction(19)
I[10:14:56.300] --> reply:textDocument/codeAction(19) 0 ms
I[10:14:56.300] --> textDocument/clangd.fileStatus

**Comment by sam-mccall** [comment]
Your compile command has `-I/usr/include/c++/7` - can you confirm the standard library headers are actually there?

(Clangd 9 is quite old and it might be worth trying a [recent snapshot](https://github.com/clangd/clangd/releases/tag/12.0.0-rc2) but I don't know of any specific reason this wouldn't work...)

**Comment by Ava-alt** [comment]
the problem has been solved, thank you very much!

**Comment by Ava-alt** [comment]
I simply switch to a newer version of Clangd，i Confirmed  the include path in the compile_commands.json  war right.    i think i can  share the method  if  anyone stumbling upon this problem in the future。

**Comment by Ava-alt** [comment]
There seems to be a problem about semantic highlighting。 Hope to give some guidance
![image](https://user-images.githubusercontent.com/79972211/110432957-d047c880-80ea-11eb-85ed-00d5ec5cc5e6.png)
![image](https://user-images.githubusercontent.com/79972211/110433069-f66d6880-80ea-11eb-886f-e8fcbb9af3a9.png)
